shared/js/background.js

```diff
@@ -242,21 +246,44 @@ chrome.webRequest.onBeforeRequest.addListener(
             return
         }
 
+        if (thisTab.failedUpgradeUrls[requestData.url]) {
+            console.log('already tried https upgrades for this URL and failed, skip:\n' + requestData.url)
+            return
+        }
+
         // Avoid redirect loops
         if (thisTab.httpsRedirects[requestData.requestId] >= 7) {
             console.log('HTTPS: cancel https upgrade. redirect limit exceeded for url: \n' + requestData.url)
+
             return {redirectUrl: thisTab.downgradeHttpsUpgradeRequest(requestData)}
         }
 
         // Is this request from the tab's main frame?
         const isMainFrame = requestData.type === 'main_frame' ? true : false
 
+        if (isMainFrame &&
+                thisTab.lastHttpsUpgrade &&
+                thisTab.lastHttpsUpgrade.url === requestData.url &&
+                Date.now() - thisTab.lastHttpsUpgrade.time < 1000) {
+
+            console.log('already tried upgrading this url on this tab a few moments ago ' +
+                'and it didn\'t complete successfully, abort:\n' +
+                requestData.url)
+            thisTab.downgradeHttpsUpgradeRequest(requestData)
+            return
+        }
```

Comment:
The time-based approach feels weird, but it's the only thing that worked for the RaisingChildren case.

One thing I tried was to cache the URL and clear it after successful page load like we do on the Safari branch. However, on Chrome the redirect page was still counted as a successful page load, so the URL was cleared prematurely.

Reply:
yea, the timer doesn't seem ideal. Is there a "page successfully loaded and isn't going to be redirected anymore" event where we could clear `thisTab.lastHttpsUpgrade`?

Reply:
What about clearing `lastHttpsUpgrade` here? https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/master/shared/js/tabManager.js#L139

Reply:
> Is there a "page successfully loaded and isn't going to be redirected anymore" event where we could clear `thisTab.lastHttpsUpgrade`

Not sure. For Chrome, the JS redirect page I [mentioned above](https://github.com/duckduckgo/duckduckgo-privacy-extension/pull/83#discussion_r163517171) *is* the successfully loaded page. We can't tell that the page isn't going to do any more redirects unless we:

- parse the content (this way madness lies)
- wait a bit (what we're doing now)

> What about clearing `lastHttpsUpgrade` here?

Can you elaborate? The headers will come before the redirect page content is parsed and acted upon, right?

Reply:
I see, thanks, that makes more sense now. What about this `chrome.webNavigation` API? Does it do the redirect before onCompleted fires there?

Reply:
https://developer.chrome.com/extensions/webNavigation

Reply:
Nice find, but `onCompleted` seems to fire before the page loads is complete.

Reply:
weird, really? It says in the docs it's after DOMCompleted.

Reply:
Here's what I did. Testing on Chrome:

1. Add the following to [this line or thereabouts](https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/eafef4fc7c47484f6be311a165d9863ef73ded34/shared/js/background.js#L70):
    
    ```
    chrome.webNavigation.onCompleted.addListener(function (details) {
       const tab = tabManager.get(details)
       console.log("navigation completed for: " + tab.url)
       delete tab.lastHttpsUpgrade
    })
    ```
2. Open http://raisingchildren.net.au/articles/safe_water_temperature.html - it'll enter a redirect loop because `tab.lastHttpsUpgrade` isn't there anymore
3. Open the debug log for the extension, here's what I see:

![screenshot from 2018-01-24 17-46-56](https://user-images.githubusercontent.com/1431577/35348146-15423324-012f-11e8-8c5c-7b065b921624.png)

That `tab.status` log line appears in the `tabs.onUpdated` event, [here](https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/eafef4fc7c47484f6be311a165d9863ef73ded34/shared/js/tabManager.js#L112). I've tried removing `lastHttpsUpgrade` in that event, and it's still way too early.